File: leetcode-scala/src/dfs/medium/ValidateBinarySearchTree.scala
package dfs.medium

import tools.TreeNode

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/7/15 00:02
 * @Description ValidateBinarySearchTree
 */
object ValidateBinarySearchTree {

    /**
     * 660 ms   94.74%
     * 52.7 MB  100.00%
     * 使用Long是因为测试用例卡Integer的边界值
     * @param root
     * @return
     */
    def isValidBST(root: TreeNode): Boolean = {

        def validate(node: TreeNode, min: Long, max: Long): Boolean = {
            if (node == null) return true
            if (node.value <= min || node.value >= max) return false
            validate(node.left, min, node.value) && validate(node.right, node.value, max)
        }

        validate(root, Long.MinValue, Long.MaxValue)
    }

    /**
     * 搜索二叉树的中序遍历是升序
     * leetcode执行代码是正确的 但是提交过不了 本地结果是正确的
     * @param root
     * @return
     */
    var pre: Long = Long.MinValue
    def isValidBST2(root: TreeNode): Boolean = {
        if (root == null) return true

        def validate(node: TreeNode): Boolean = {
            if (pre >= node.value) return false
            pre = node.value.toLong
            true
        }

        isValidBST2(root.left) && validate(root) && isValidBST2(root.right)
    }

    def main(args: Array[String]): Unit = {
        val root = new TreeNode(0)
        print(isValidBST2(root))
    }

}

File: leetcode-java/src/classify/recursion/ValidateBinarySearchTree.java
package classify.recursion;

import tools.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author yutiantang
 * @create 2021/4/6 22:43
 */
public class ValidateBinarySearchTree {


    /**
     * 中序遍历
     * 2ms      27.69%
     * 38.3MB   11.01%
     * @param root
     * @return
     */
    private final List<Integer> values = new ArrayList<>();
    public boolean isValidBST(TreeNode root) {
        midTraversal(root);
        if (values.size() < 2) {
            return true;
        }
        long pre = Long.MIN_VALUE;
        for (int value : values) {
            if (pre >= value) {
                return false;
            }
            pre = value;
        }

        return true;
    }

    private void midTraversal(TreeNode node) {
        if (node.left != null) {
            midTraversal(node.left);
        }
        values.add(node.val);
        if (node.right != null) {
            midTraversal(node.right);
        }
    }

    /**
     * 官方题解
     * 0ms      100.00%
     * 37.8MB   95.43%
     */
    public boolean isValidBST2(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    public boolean isValidBST(TreeNode node, long lower, long upper) {
        if (node == null) {
            return true;
        }

        if (node.val <= lower || node.val >= upper) {
            return false;
        }

        return isValidBST(node.left, lower, node.val) && isValidBST(node.right, node.val, upper);
    }
}
